hyperstart/api/grpc/hyperstart.proto

```diff
@@ -30,34 +32,42 @@ service HyperstartService {
 	rpc OnlineCPUMem(OnlineCPUMemRequest) returns (google.protobuf.Empty);
 }
 
-message AddContainerRequest {
+message CreateContainerRequest {
 	Container container = 1;
 	Process init = 2;
+	// config is a marshalled OCI configuration JSON file (a.k.a. config.json).
+	// This is optional and can be used by agents implementing the hyperstart protocol to apply
+	// any resource or security constraints defined by this configuration.
+	string config = 3;
```

Comment:
I prefer defining the desired fields in the message type Container&Process.
I don't want an embedded string with another language(json).
We might not use the OCI configuration JSON in the hyperstart since we will use libcontainer directly.

Reply:
How about simply putting the OCI specs.Spec and specs.Process in place (with named fields instead of dummy string)? libcontainer in the new agent can use them directly and we do not have to modify hyperstart protocol every time when we want to add new supported fields. Of course it would add a dependency on OCI spec versions but I think it is OK since we want to be OCI compliant anyway.

Reply:
@laijs @bergwolf Defining or copying the right field would be ideal, but that means copying most of the OCI `specs.Specs` in this file as there is no OCI specification gRPC proto file.
If that is what you want, I'll create an oci.proto separate file to not pollute the hyperstart.proto one.

Reply:
oci.proto sounds good. But it should be along with a verification code(or tool) to prove that the oci.proto and the imported OCI specs are fully matched.

Reply:
@laijs @bergwolf I'm implementing `oci.proto` and because of gRPC Go conversion assumptions, we will have to handle 2 different structures:

The gRPC generated Go structure:
```Go
type Spec struct {
	Version     string            `protobuf:"bytes,1,opt,name=Version" json:"Version,omitempty"`
	Process     *OCIProcess       `protobuf:"bytes,2,opt,name=Process" json:"Process,omitempty"`
	Root        *Root             `protobuf:"bytes,3,opt,name=Root" json:"Root,omitempty"`
	Hostname    string            `protobuf:"bytes,4,opt,name=Hostname" json:"Hostname,omitempty"`
	Mounts      []*Mount          `protobuf:"bytes,5,rep,name=Mounts" json:"Mounts,omitempty"`
	Hooks       *Hooks            `protobuf:"bytes,6,opt,name=Hooks" json:"Hooks,omitempty"`
	Annotations map[string]string `protobuf:"bytes,7,rep,name=Annotations" json:"Annotations,omitempty" protobuf_key:"bytes,1,opt,name=key" protobuf_val:"bytes,2,opt,name=value"`
	Linux       *Linux            `protobuf:"bytes,8,opt,name=Linux" json:"Linux,omitempty"`
}
```

The OCI Spec Go structure:
```Go
type Spec struct {
	Version string `json:"ociVersion"`
	Process *Process `json:"process,omitempty"`
	Root Root `json:"root"`
	Hostname string `json:"hostname,omitempty"`
	Mounts []Mount `json:"mounts,omitempty"`
	Hooks *Hooks `json:"hooks,omitempty" platform:"linux,solaris"`
	Annotations map[string]string `json:"annotations,omitempty"`
	Linux *Linux `json:"linux,omitempty" platform:"linux"`
	Solaris *Solaris `json:"solaris,omitempty" platform:"solaris"`
	Windows *Windows `json:"windows,omitempty" platform:"windows"`
}
```

As you can see those are different structures and we will have to go through the following steps:

```
runtime (host) -> OCI spec structure  ---- conversion ----> gRPC spec structure -> gRPC binary -> gRPC structure --- conversion ----> OCI spec structure -> agent (guest)  
```

So by not tunnelling a marshalled OCI spec through gRPC, we're adding 2 additional conversion layers: from OCI to gRPC (on the host) and then from gRPC to OCI (on the guest).

Are we all ok with that ?

Reply:
cc @jodh-intel

Reply:
cc @laijs @bergwolf Are we all good with the above approach?

Reply:
@sameo can you please share your oci.proto? I did some preliminary test and grpc-go did generate expected spec struct w.r.t. the `Solaris *Solaris` field. So I want to understand the issue first.

PS, my test is [here](https://gist.github.com/bergwolf/c1dab2956a422dac4831d5dd9e3c3587)

Reply:
@bergwolf Here is mine: https://gist.github.com/sameo/d8dc8c8720a67ae92100fc6d75f78987

For all non scalar fields, `gRPC` will generate a pointer to the struct. So if you create the `Mount` message in your test, you will see that `Spec` becomes:
```Go
type Spec struct {
	Version     string            `protobuf:"bytes,1,opt,name=Version" json:"Version,omitempty"`
	Process     *OCIProcess       `protobuf:"bytes,2,opt,name=Process" json:"Process,omitempty"`
	Root        *Root             `protobuf:"bytes,3,opt,name=Root" json:"Root,omitempty"`
	Hostname    string            `protobuf:"bytes,4,opt,name=Hostname" json:"Hostname,omitempty"`
	Mounts      []*Mount          `protobuf:"bytes,5,rep,name=Mounts" json:"Mounts,omitempty"`
	Hooks       *Hooks            `protobuf:"bytes,6,opt,name=Hooks" json:"Hooks,omitempty"`
	Annotations map[string]string `protobuf:"bytes,7,rep,name=Annotations" json:"Annotations,omitempty" protobuf_key:"bytes,1,opt,name=key" protobuf_val:"bytes,2,opt,name=value"`
	Linux       *Linux            `protobuf:"bytes,8,opt,name=Linux" json:"Linux,omitempty"`
}
```

and for example you get `Mounts      []*Mount` as opposed to `Mounts      []Mount` for the OCI spec.

Reply:
@laijs @bergwolf What about something more structured than my initial proposal:

```
diff --git a/hyperstart/api/grpc/hyperstart.proto b/hyperstart/api/grpc/hyperstart.proto
index 7a062ad..1691d14 100644
--- a/hyperstart/api/grpc/hyperstart.proto
+++ b/hyperstart/api/grpc/hyperstart.proto
@@ -35,10 +35,21 @@ service HyperstartService {
 message CreateContainerRequest {
        Container container = 1;
        Process init = 2;
-       // config is a marshalled OCI configuration JSON file (a.k.a. config.json).
-       // This is optional and can be used by agents implementing the hyperstart protocol to apply
-       // any resource or security constraints defined by this configuration.
-       string config = 3;
+
+       // spec is the container specification.
+       // With OCI its payload will be a marshalled config.json.
+       Spec spec = 3;
+}
+
+message Spec {
+       // type is the specification type. For now we only support the "OCI" type.
+       string type = 1;
+
+       // version is the specification version
+       string version = 2;
+
+       // payload is the marshalled specification payload, e.g. config.json.
+       string payload = 3;
 }
 
 message StartContainerRequest {
```